daemon/daemon.go

```diff
@@ -1032,23 +1036,56 @@ func (daemon *Daemon) TagImage(newTag reference.Named, imageName string, force b
 	return nil
 }
 
+func writeDistributionProgress(cancelFunc func(), outStream io.Writer, progressChan <-chan xfer.Progress) {
+	sf := streamformatter.NewJSONStreamFormatter()
+	for prog := range progressChan {
+		if prog.Message != "" {
+			formatted := sf.FormatStatus(prog.ID, prog.Message)
+			if _, err := outStream.Write(formatted); err != nil {
+				cancelFunc()
```

Comment:
Cancelling the context causes an error to be logged by the pull code:

    DEBU[1239] Error trying v2 registry: context canceled

If we logged an error here, it might end up in the log several times, because many progress messages could be in flight during the cancellation. So I think we would need to make sure to only log once.